src/compiler/core.ts

```diff
@@ -2552,4 +2561,168 @@ namespace ts {
+            return some(entries, file => fileNameEqual(file, name));
+        }
+
+        function updateFileSystemEntry(entries: ReadonlyArray<string>, baseName: string, isValid: boolean) {
+            if (hasEntry(entries, baseName)) {
+                if (!isValid) {
+                    return filter(entries, entry => !fileNameEqual(entry, baseName));
+                }
+            }
+            else if (isValid) {
+                return entries.concat(baseName);
+            }
+            return entries;
+        }
+
+        function writeFile(fileName: string, data: string, writeByteOrderMark?: boolean): void {
+            const path = toPath(fileName);
+            const result = cachedReadDirectoryResult.get(getDirectoryPath(path));
+            const baseFileName = getBaseFileName(normalizePath(fileName));
+            if (result) {
+                result.files = updateFileSystemEntry(result.files, baseFileName, /*isValid*/ true);
+            }
+            return host.writeFile(fileName, data, writeByteOrderMark);
+        }
+
+        function fileExists(fileName: string): boolean {
+            const path = toPath(fileName);
+            const result = cachedReadDirectoryResult.get(getDirectoryPath(path));
+            const baseName = getBaseFileName(normalizePath(fileName));
+            return (result && hasEntry(result.files, baseName)) || host.fileExists(fileName);
+        }
+
+        function directoryExists(dirPath: string): boolean {
+            const path = toPath(dirPath);
+            return cachedReadDirectoryResult.has(path) || host.directoryExists(dirPath);
+        }
+
+        function createDirectory(dirPath: string) {
+            const path = toPath(dirPath);
+            const result = cachedReadDirectoryResult.get(getDirectoryPath(path));
+            const baseFileName = getBaseFileName(path);
+            if (result) {
+                result.directories = updateFileSystemEntry(result.directories, baseFileName, /*isValid*/ true);
+            }
+            host.createDirectory(dirPath);
+        }
+
+        function getDirectories(rootDir: string): string[] {
+            if (canWorkWithCacheForDir(rootDir)) {
+                return getFileSystemEntries(rootDir).directories.slice();
+            }
+            return host.getDirectories(rootDir);
+        }
+        function readDirectory(rootDir: string, extensions?: ReadonlyArray<string>, excludes?: ReadonlyArray<string>, includes?: ReadonlyArray<string>, depth?: number): string[] {
+            if (canWorkWithCacheForDir(rootDir)) {
+                return matchFiles(rootDir, extensions, excludes, includes, host.useCaseSensitiveFileNames, getCurrentDirectory(), depth, path => getFileSystemEntries(path));
+            }
+            return host.readDirectory(rootDir, extensions, excludes, includes, depth);
+        }
+
+        function addOrDeleteFileOrFolder(fileOrFolder: string) {

```

Comment:
No because caller to this function is normally from the watch and directory watches dont specify what the action was. For now this is best one can do to infer